nrf/boards: Add support for nina_evk_b3 board

bruno963852 opened the pull request:
Hello, friends, first time contributor here, but long time user!

It's also my first time contributing to a large open source project like this, sou i hope a did the job right.

I'm currently using the U-blox NINA MCUs (B1 and b3) at work, so i made some changes in the port to add support do the B3 board and enable all the features that i need. I decided to start the PR with just adding support for the board, because there's no conflicts, and add the other changes later as i get used to this.

Added support for the NINA-EVK-B3 board, with the correct pin assignment.

[edit]
I've also enabled the @viper and native code emitters on the nrf port. It's working like a charm on the evk_b3 board that is an nrf52840. Also enalbed the direct memory access functionality to the machine module, for when you want that ultra high performance.

Also enabled the errno module for working better with the Pymakr extension on VSCode.

bruno963852 committed:
nrf/boards: Add support for nina_evk_b3 board

mcauser commented:
The dev boards:
https://www.u-blox.com/en/product/evk-nina-b1
https://www.u-blox.com/en/product/evk-nina-b3

bruno963852 committed:
nrf: Added support for @viper and @native. added machine.memxx

Enabled the @viper and @native emitters on mpconfigport.h.
Enabled errno module on mpconfigport.h to work better with the pymakr extension on vscode.
enabled machine.memxx for direct memory access.

bruno963852 committed:
nrf: Added support for viper and native. added machine.memxx

dpgeorge commented:
Thanks for the contribution!  I assume you tested this on hardware?

dpgeorge commented on ports/nrf/boards/evk_nina_b3/mpconfigboard.h:
```diff
@@ -0,0 +1,84 @@
+//Boar data
+#define MICROPY_HW_BOARD_NAME       "EVK_NINA_B3"	//Board name
+#define MICROPY_HW_MCU_NAME         "NRF52840"	//Microcontroller name
+#define MICROPY_PY_SYS_PLATFORM     "nrf52840"	//Returns when you call sys.ṕlatform
```
this would be better as "nrf52" so that all nrf52 boards can be treated the same way with code (and the MCU can be retrieved via other means)

dpgeorge commented on ports/nrf/boards/evk_nina_b3/mpconfigboard.h:
```diff
@@ -0,0 +1,84 @@
+
+// #define MICROPY_EMIT_ARM	     (1)
+//#define MICROPY_PY_ARRAY_SLICE_ASSIGN (1)
+//#define MICROPY_PY_SYS_STDFILES     (1)
```
please remove unused comment lines like these

bruno963852 replied:
my bad... forgot to delete...

dpgeorge commented on ports/nrf/mpconfigport.h:
```diff
@@ -44,8 +44,8 @@
-#define MICROPY_EMIT_THUMB          (0)
-#define MICROPY_EMIT_INLINE_THUMB   (0)
+#define MICROPY_EMIT_THUMB          (1)
+#define MICROPY_EMIT_INLINE_THUMB   (1)
```
These should not be enabled for all boards by default, it's better to just define them in `mpconfigboard.h` if that board has enough space (like nrf52840).

Since the defaults for these are "disabled" the best way forward here would be to remove these 2 lines from this file, and enable the feature in `mpconfigboard.h`.

bruno963852 replied:
Yeah, i really though so....
So, should i remove these from the mpconfigport and add it to all the mpconfigboard of the nrf boards for not loosing compatibility?

dpgeorge replied:
Yes remove them from `mpconfigport.h` (this won't change anything because disabled is the default), and then enable (add) only to this new board that you define in this PR.

dpgeorge commented on ports/nrf/mpconfigport.h:
```diff
@@ -123,15 +123,15 @@
 #define MICROPY_PY_IO               (0)
 #define MICROPY_PY_IO_FILEIO        (0)
-#define MICROPY_PY_UERRNO           (0)
+#define MICROPY_PY_UERRNO           (1)
```
Again, since some boards like the microbit don't have much space, these should be defined per-board rather than for all boards here.

dpgeorge commented on ports/nrf/mpconfigport.h:
```diff
@@ -123,15 +123,15 @@
 #define MICROPY_PY_URE              (0)
 #define MICROPY_PY_UHEAPQ           (0)
-#define MICROPY_PY_UHASHLIB         (0)
+#define MICROPY_PY_UHASHLIB         (1)
```
same as above

bruno963852 committed:
Merge remote-tracking branch 'origin/nrf_improvements' into nrf_improvements

bruno963852 commented:
> Thanks for the contribution! I assume you tested this on hardware?

you assumed correctly. I believe i made all the changes that were requested.

dpgeorge commented:
Thanks for updating.  I reworked the commits a little to have a clear distinction between adding `machine.memXX` and adding the new board.  Merged in 94fb5e7f5a409e9596657508c6a9ed1c609de267 and ea2d83e96175ef1ed260cca917024502e01a5546
